Modify to a contiguous memory map file using multiple files

MichaelKim20 opened the pull request:
A memory map file consists of several files.
Previously, there was a gap between the files. We got rid of it and connected it as one.
And when we save a block or read a block, we can serialize it from the inside.
We can expect better performance because we can reduce memory replication.
![MMFile](https://user-images.githubusercontent.com/11639939/65366956-8dc7cd00-dc65-11e9-950b-41217da92fe7.gif)

AndrejMitrovic commented on source/agora/common/BlockStorage.d:
```diff
@@ -382,6 +397,50 @@ public class BlockStorage
+        auto finds
+            = this.merkle_root_idx[].find!((a, b) => a.hash == b)(hash_bytes);
+
+        if (!finds.empty)
```
It's better to return early. So:

```D
if (finds.empty)
    return false;
```

And then the rest of the code can be unindented by one level. It makes it easier to reason about code this way.

MichaelKim20 replied:
Yes, I'll modify it.

Geod24 commented on source/agora/common/BlockStorage.d:
```diff
@@ -781,5 +781,5 @@ version(none) unittest
     }
 
-    other.release();
+    other.unMap();
```
Sorry, I noticed that after. I thought `release` was unused and so removed it. But it is a better name, all things considered, so the integration test PR fixes this.

Geod24 commented:
You should rebase on v0.x.x

MichaelKim20 commented:
A memory map file consists of several files.
Previously, there was a gap between the files. We got rid of it and connected it as one.
And when we save a block or read a block, we can serialize it from the inside.
We can expect better performance because we can reduce memory replication.

MichaelKim20 commented:
Std.mmfile was modified to obtain the address of the memory map file.
However, it was difficult to secure adjacent memory continuously.

So whenever a new memory map is created. 
It was possible to change the memory address of the base point.

However, I thought it was not appropriate to have to modify std.mmfile.
This was resolved by the concept of position of the file.

Then, when accessing the portion at the edge of the file in the memory map, it was done one by one.

I think I've met all the requirements now.

Geod24 commented on source/agora/common/BlockStorage.d:
```diff
@@ -33,17 +33,8 @@ import std.path;
-private immutable size_t MFILE_RESERVE_SIZE = 64 * 1024;
-
-/// The maximum number of block in one file
-private immutable ulong MFILE_MAX_BLOCK = 100;
```
In this commit (the second) those constants are removed, not renamed, so this commit does not compile.

MichaelKim20 replied:
fixed

MichaelKim20 replied:
fixed

Geod24 commented on source/agora/common/BlockStorage.d:
```diff
@@ -113,73 +113,37 @@ public class BlockStorage
+            this.file_base = MapSize * this.file_index;
         }
-        else
+        catch (Exception ex)
```
So the failure does not get reported to the caller. Does that mean we are okay with this component failing ?

MichaelKim20 replied:
fixed

MichaelKim20 replied:
fixed

Geod24 commented on source/agora/common/BlockStorage.d:
```diff
@@ -149,19 +149,24 @@ public class BlockStorage
-        destroy(this.file);
-        GC.free(&this.file);
-        this.file = null;
+        try
```
Did you see `release` throw ? Because it should never happen.

MichaelKim20 replied:
fixed

Geod24 commented on source/agora/common/BlockStorage.d:
```diff
@@ -244,6 +244,7 @@ public class BlockStorage
 
         return true;
     }
+
```
Unrelated change (commit 926779701294e16eba7c744344ad30f976d02033)

Geod24 commented on source/agora/common/BlockStorage.d:
```diff
@@ -359,12 +360,33 @@ public class BlockStorage
+                const size_t x1 = to - this.file_base;
+                data = cast(ubyte[])this.file[x0 .. x1];
+            }
+            catch (Exception ex)
```
Flat-out ignoring an exception does not seem like the best way to go

MichaelKim20 replied:
fixed

Geod24 commented on source/agora/common/BlockStorage.d:
```diff
@@ -377,13 +399,76 @@ public class BlockStorage
+                    this.writeByte(pos + idx, e);
+            }
+        }
+        catch (Exception ex)
```
Same here, why are we ignoring Exception ?

MichaelKim20 replied:
fixed

MichaelKim20 replied:
fixed

Geod24 reviewed (changes requested):
Good commit separation. The review was much simpler. Thanks for that 👍 
My recommendation: The commit title should indicate what & why something is being done. A commit message like "Modify function read, write" is not informative. "Remove ability to read unmapped file" gives much more info, for example (that's just an example). However, things like "Remove unused variable xxx" or "Fix typo in documentation", for example, are fine because it is self evident from the diff.
Last thing, there have been a lot of empty `try` / `catch` added. They seem to have been added to make the function `nothrow`, however that means that errors are now ignored, and we cannot guarantee the integrity of the data on disk, which is pretty bad.

Geod24 commented:
Can you submit the first 4 commits in an independent PR ? They are good to go.

Geod24 commented:
@MukeunKim :
Please do not ignore `Exception`.
The pattern:
```d
try { /* ... */ }
catch (Exception e) {}
```
Is very bad because if an exception is triggered, we won't know.
I understand that is done because there are some `nothrow` in place.
In this case, for the time being, I recommend to use:
```d
void func () nothrow
{
    scope (failure) assert(0);
    // ...
}
```

This will make the function `nothrow`

Geod24 commented:
Ah, it looks like the last commit should be merged with the first one. But I recommend to just `assert` for the time being.

Commit:
Add function that outputs an error with the attribute `@trusted nothrow`

In 'try..catch..' inside a function with nothrow attribute
try{...} catch(Exception) { 'Use Here' }

Commit:
Changed the memory map file to use as if it were connected to memory

Unused areas between maps have been removed.

Geod24 commented:
Ready for a new review

AndrejMitrovic commented:
What about the test-case including multiple transactions in the block?

Commit:
Add test-case of `BlockStorage` with multiple transactions in the block

It allows testing even in multi-transactional blocks.

MukeunKim commented:
@AndrejMitrovic

MichaelKim20 commented:
Added test casings for multi-transaction to block.
I have also verified normal operation for large blocks.
I'm trying to add @AndrejMitrovic  to coworker for the test code, but it doesn't work.

AndrejMitrovic commented:
> I'm trying to add @AndrejMitrovic to coworker for the test code, but it doesn't work.

It doesn't matter, I just want us to have a test.

AndrejMitrovic commented on source/agora/common/BlockStorage.d:
```diff
@@ -301,8 +280,17 @@ public class BlockStorage
 
-        this.readBlockAtPosition(block, finds.front.position);
-        return true;
+        bool success = true;
```
In general, it is much better to assign failure by default, not success.

However there is no need for this boolean. This can be rewritten to:

```D
try
{
    this.readBlockAtPosition(block, finds.front.position);
    return true;
}
catch (Exception ex)
{
    this.writeLog("BlockStorage.readBlock: ", ex);
    return false;
}
```

AndrejMitrovic commented on source/agora/common/BlockStorage.d:
```diff
@@ -328,26 +316,33 @@ public class BlockStorage
+        bool success = true;
+        try
+        {
+            this.readBlockAtPosition(block, finds.front.position);
```
Same comment here as above.

AndrejMitrovic commented on source/agora/common/BlockStorage.d:
```diff
@@ -441,30 +457,79 @@ public class BlockStorage
+        }
+        catch (Exception ex)
+        {
+            writeLog("BlockStorage.readBytes: ", ex);
```
So in case of a failure, it will return a ubyte with the value 0 and the error will be ignored?

AndrejMitrovic commented on source/agora/common/BlockStorage.d:
```diff
@@ -441,30 +457,79 @@ public class BlockStorage
+        }
+        catch (Exception ex)
+        {
+            this.writeLog("BlockStorage.writeByte: ", ex);
```
Why is it ok to fail to write and then ignore the error? If the node can't write to disk, wouldn't it be better to abort the whole process and shut down the node? The data that was written will be invalid when its being read later, because some of the data is missing (the checksum will fail, once it's implemented).

AndrejMitrovic commented on source/agora/common/BlockStorage.d:
```diff
@@ -441,30 +457,79 @@ public class BlockStorage
+        }
+        catch (Exception ex)
+        {
+            this.writeLog("BlockStorage.write: ", ex);
```
Also same comment about ignoring write errors.

MichaelKim20 commented:
Review Please!

AndrejMitrovic commented:
It looks great now. How about adding `nothrow` everywhere where there a function returns a status code?

AndrejMitrovic commented:
@MukeunKim if you get too many errors trying to make them `nothrow`, or if it's too much work, then leave it the way it is. We can add `nothrow` later.

MukeunKim commented:
Ready for a new review

MichaelKim20 commented:
@Geod24 Ready for a new review ^^

Commit:
If no folder exists to store block data, it was processed to create it.

Geod24 commented on source/agora/common/BlockStorage.d:
```diff
@@ -119,79 +116,44 @@ public class BlockStorage
     ***************************************************************************/
 
-    private void map (size_t findex, size_t resize = 0) @trusted
+    private bool map (size_t findex) @trusted nothrow
```
Missing `Returns:`

Geod24 reviewed (approved):
LGTM
